lib/common/strings.c

```diff
@@ -363,52 +417,74 @@ pcmk__guint_from_hash(GHashTable *table, const char *key, guint default_val,
 long long
 crm_get_msec(const char *input)
 {
-    const char *num_start = NULL;
-    const char *units;
+    char *buf = NULL;
+    char *units = NULL; // Do not free; will point to part of buf
     long long multiplier = 1000;
     long long divisor = 1;
     long long msec = PCMK__PARSE_INT_DEFAULT;
-    size_t num_len = 0;
-    char *end_text = NULL;
 
     if (input == NULL) {
         return PCMK__PARSE_INT_DEFAULT;
     }
 
-    num_start = input + strspn(input, WHITESPACE);
-    num_len = strspn(num_start, NUMCHARS);
-    if (num_len < 1) {
+    buf = calloc(strlen(input) + 1, sizeof(char));
```

Comment:
crm_get_msec() is used often enough that I'd like to avoid an extra memory allocation. I don't see any real advantage in the clean-up function

Reply:
I have a hard time seeing repeated allocation of a few bytes as a problem, knowing how much memory we allocate -- and how often -- for XML creation and manipulation. But I don't feel like benchmarking it, so I can agree to disagree based on speculation.

A fixed-size array is no good because we could have an arbitrary amount of whitespace or an arbitrary number of digits after a decimal, and also (to be overly technical) because a long long is only guaranteed to be *at least* 64 bits.

I thought the cleanup function made string processing easier to reason about and to possibly improve or extend later. For example, we can get `units` directly from `scan_ll()`, instead of passing a useless `end_text` argument to it and getting the units via explicit two `strspn()` calls and some pointer arithmetic. More generally, it abstracts dealing with garbage characters away from the main logic: preprocess the string to get rid of the characters we don't care about, and then process the clean string for meaning.

---

It's obviously not necessary and I can do something that doesn't involve dynamic allocation.

Reply:
Eh, new version turned out better anyway for current needs.